lib/ruby_smb/client.rb

```diff
@@ -335,34 +399,97 @@ def logoff!
     #
     # @param packet [RubySMB::GenericPacket] the request to be sent
     # @return [String] the raw response data received
-    def send_recv(packet)
-      case packet.packet_smb_version
+    def send_recv(packet, encrypt: false)
+      version = packet.packet_smb_version
+      case version
       when 'SMB1'
         packet.smb_header.uid = user_id if user_id
         packet = smb1_sign(packet)
       when 'SMB2'
         packet = increment_smb_message_id(packet)
         packet.smb2_header.session_id = session_id
         unless packet.is_a?(RubySMB::SMB2::Packet::SessionSetupRequest)
-          packet = smb2_sign(packet)
+          if self.smb2
+            packet = smb2_sign(packet)
+          elsif self.smb3
+            packet = smb3_sign(packet)
+          end
         end
       else
         packet = packet
       end
-      dispatcher.send_packet(packet)
-      raw_response = dispatcher.recv_packet
+
+      if [RubySMB::SMB2::Packet::SessionSetupRequest, RubySMB::SMB2::Packet::NegotiateRequest].none? {|klass| packet.is_a? klass} &&
+          ["0x0300", "0x0302", "0x0311"].include?(@dialect) &&
+          (@encryption_required || encrypt)
+        send_encrypt(packet)
+        raw_response = recv_encrypt
+        loop do
```

Comment:
The response can be an asynchronous operation, which means the server needs more time to process locally (e.g. write/read file operation). It will first reply with a `STATUS_PENDING` response packet and finally send with the expected response after it's done processing locally. When receiving such asynchronous response, the client needs to read the socket again to get the actual response.

This implementation is far from perfect, but adding a `sleep 1` help to ensure the client will get the expected response the second time it reads the socket. Without it, it might timeout since the socket `read_timeout` is set to only 30 sec. by default. Maybe I'm wrong and it is not necessary, I will do some tests.

More info about this:
- [3.3.4.2 Sending an Interim Response for an Asynchronous Operation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/c1692ece-7c6d-45c8-a0e3-6b53c0483030)
- [3.2.5.1.5 Handling Asynchronous Responses](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/a01c6101-b94b-45c1-b7ec-cbcc94d51138)